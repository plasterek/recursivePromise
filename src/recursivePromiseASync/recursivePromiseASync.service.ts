export async function recursivePromiseASync<T>(promisesArray: Array<() => Promise<T>>): Promise<Array<T>> {
  if (promisesArray.length === 0) {
    return [];
  } else {
    const [first, ...rest]: (() => Promise<T>)[] = promisesArray;
    const result: Awaited<T> = await first()
      .then((data) => data)
      .catch((err) => err);
    const remaining: T[] = await recursivePromiseASync(rest);
    return [result, ...remaining];
  }
}
