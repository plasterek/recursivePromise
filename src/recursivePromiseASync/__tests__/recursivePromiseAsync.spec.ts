import { recursivePromiseASync } from "../recursivePromiseASync.service";

describe("RecursivePromise function", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  const promise1: () => Promise<unknown> = () => Promise.resolve("resolved1");
  const promise2: () => Promise<unknown> = () => Promise.resolve("resolved2");
  const promise3: () => Promise<unknown> = () => Promise.resolve("resolved3");
  const promise4: () => Promise<unknown> = () => Promise.resolve("resolved4");
  const promise5: () => Promise<unknown> = () => Promise.reject("rejected5");
  const promise6: () => Promise<unknown> = () => Promise.resolve("resolved6");
  const promise7: () => Promise<unknown> = () => Promise.resolve("resolved7");
  const arrayOfPromises: (() => Promise<unknown>)[] = [
    promise1,
    promise2,
    promise3,
    promise4,
    promise5,
    promise6,
    promise7,
  ];

  describe("When given array of promises is empty", () => {
    it("It returns Promise with with empty array", async () => {
      //given
      const array: any[] = [];
      const result: unknown[] = [];
      // when
      await recursivePromiseASync(array).then((data) => result.push(data));
      //then
      expect.assertions(1);
      expect(result).toMatchSnapshot();
    });
  });
  describe("When array of promises is given", () => {
    it("It returns Promise with array with same lenght as given array", async () => {
      //given
      let resultArrayLenght: number = 0;
      //when
      const arrayOfPromisesLenght: number = arrayOfPromises.length;
      await recursivePromiseASync(arrayOfPromises).then((array) => {
        resultArrayLenght = array.length;
      });
      //then
      expect.assertions(1);
      expect(resultArrayLenght).toBe(arrayOfPromisesLenght);
    });
  });
  describe("When array of promises is given", () => {
    it("It return Promise with array of resolved and rejected promises results", async () => {
      //given
      let result: unknown[] = [];
      //when
      await recursivePromiseASync(arrayOfPromises).then((data) => (result = data));
      //then
      expect.assertions(1);
      expect(result).toMatchSnapshot();
    });
  });
});
