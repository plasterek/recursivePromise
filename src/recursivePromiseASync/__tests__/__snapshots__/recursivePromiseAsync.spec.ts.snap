// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecursivePromise function When array of promises is given It return Promise with array of resolved and rejected promises results 1`] = `
[
  "resolved1",
  "resolved2",
  "resolved3",
  "resolved4",
  "rejected5",
  "resolved6",
  "resolved7",
]
`;

exports[`RecursivePromise function When given array of promises is empty It returns Promise with with empty array 1`] = `
[
  [],
]
`;
